okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java

```diff
@@ -0,0 +1,163 @@
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        ConcurrencyLimitTag tagState = chain.request().tag(ConcurrencyLimitTag.class);
+        tagState.invalidate();
+        tagState.setListener(limiters.limiter(chain.request()));
+        return chain.proceed(chain.request());
+    }
+
+    public static Callback wrapCallback(Callback callback) {
+        return new Callback() {
+            @Override
+            public void onFailure(Call call, IOException e) {
+                Optional.ofNullable(call.request().tag(ConcurrencyLimitTag.class)).ifPresent(ConcurrencyLimitTag::invalidate);
+                callback.onFailure(call, e);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                Response newResponse =
+                        Optional.ofNullable(call.request().tag(ConcurrencyLimitTag.class))
+                                .map(t -> wrapResponse(t, response))
+                        .orElse(response);
+                callback.onResponse(call, newResponse);
+            }
+        };
+    }
+
+    public static Request wrapRequest(Request request) {
+        return request.newBuilder().tag(ConcurrencyLimitTag.class, new ConcurrencyLimitTag()).build();
+    }
+
+    private static Response wrapResponse(ConcurrencyLimitTag tag, Response response) {
+        if (response.body() == null) {
+            return response;
+        }
+        ResponseBody currentBody = response.body();
+        ResourceDeallocator deallocator = new ResourceDeallocator(tag);
+        ResponseBody newResponseBody =
+                ResponseBody.create(currentBody.contentType(), currentBody.contentLength(),
+                        new ReleaseConcurrencyLimitBufferedSource(currentBody.source(), tag, deallocator));
+        deallocator.timeout(1, TimeUnit.MINUTES);
+        deallocator.enter();
+        return response.newBuilder()
+                .body(newResponseBody)
+                .build();
+    }
+
+    static final class ConcurrencyLimitTag {
+        private Limiter.Listener listener;
+        private boolean wasDropped = false;
+
+        private void invalidate() {
+            if (listener == null) {
+                return;
+            }
+
+            if (wasDropped) {
+                listener.onDropped();
+            } else {
+                listener.onIgnore();
+            }
+            listener = null;
+            wasDropped = false;
+        }
+
+        private void setListener(Limiter.Listener listener) {
+            checkState(listener == null);
+            this.listener = listener;
+        }
+
+        public void success() {
+            listener.onSuccess();
+        }
+
+        public void wasDropped() {
+            wasDropped = true;
+        }
+    }
+
+    private static final class ResourceDeallocator extends AsyncTimeout {
```

Comment:
don't think it's worth worrying about this case. if clients aren't releasing resources properly they're going to lock things up eventually anyway. At a minimum it should be a separate change from the concurrency limiting

Reply:
discussed more in person: going to timeout on acquiring a limit vs releasing a limit which has the added benefit of always allowing requests through after some period of time